src/server/scripts/Northrend/Naxxramas/boss_four_horsemen.cpp

```diff
@@ -82,315 +99,510 @@ enum FourHorsemen
+                    case EVENT_BERSERK:
+                        DoCastAOE(SPELL_BERSERK, true);
+                        break;
+                    case EVENT_MARK:
+                        DoCastAOE(SPELL_LADY_MARK, true);
+                        events.ScheduleEvent(EVENT_MARK, 15 * IN_MILLISECONDS);
+                        break;
+                    case EVENT_VOIDZONE:
+                        if (Unit* target = SelectTarget(SELECT_TARGET_RANDOM, 0u, 45.0f, true))
+                            DoCast(target, SPELL_VOID_ZONE, true);
+                        events.ScheduleEvent(EVENT_VOIDZONE, 15 * IN_MILLISECONDS);
+                        break;
+                }
             }
 
-            Talk(SAY_DEATH);
+            if (me->HasUnitState(UNIT_STATE_CASTING))
+                return;
+
+            if (Unit* target = SelectTarget(SELECT_TARGET_NEAREST, 0u, 45.0f, true))
+                DoCast(target, SPELL_SHADOW_BOLT);
+            else
+                DoCastAOE(SPELL_UNYIELDING_PAIN);
         }
+    };
 
-        void EnterCombat(Unit* /*who*/) override
-        {
-            _EnterCombat();
-            Talk(SAY_AGGRO);
+    CreatureAI* GetAI(Creature* creature) const override
+    {
+        return GetInstanceAI<boss_four_horsemen_ladyAI>(creature);
+    }
+};
 
-            events.ScheduleEvent(EVENT_MARK, 15000);
-            events.ScheduleEvent(EVENT_CAST, 20000 + rand32() % 5000);
-            events.ScheduleEvent(EVENT_BERSERK, 15*100*1000);
-        }
+static const Position sirPath[3] = { { 2533.141f, -2922.14f, 241.2764f }, { 2525.254f, -2905.907f, 241.2761f }, { 2517.636f, -2897.253f, 241.2758f } };
+class boss_four_horsemen_sir : public CreatureScript
+{
+    public:
+    boss_four_horsemen_sir() : CreatureScript("boss_four_horsemen_sir") { }
 
-        void UpdateAI(uint32 diff) override
+    struct boss_four_horsemen_sirAI : public boss_four_horsemen_baseAI
+    {
+        boss_four_horsemen_sirAI(Creature* creature) : boss_four_horsemen_baseAI(creature, SIR, sirPath) { }
+        void BeginFighting() override
         {
-            if (nextWP && movementStarted && !movementCompleted && !nextMovementStarted)
-            {
-                nextMovementStarted = true;
-                me->GetMotionMaster()->MovePoint(nextWP, WaypointPositions[nextWP]);
-            }
+            events.ScheduleEvent(EVENT_BERSERK, 10 * MINUTE * IN_MILLISECONDS);
+            events.ScheduleEvent(EVENT_MARK, 24 * IN_MILLISECONDS);
+            // @todo actual timing, this is for test
+            events.ScheduleEvent(EVENT_HOLYWRATH, urandms(15, 20));
+        }
 
-            if (!UpdateVictim() || !CheckInRoom() || !movementCompleted)
+        void _UpdateAI(uint32 diff) override
+        {
+            if (!me->IsInCombat())
+                return;
+            if (!_ourMovementFinished)
                 return;
+            if (me->getThreatManager().isThreatListEmpty())
+                return EnterEvadeMode();
```

Comment:
please use "return;" when the return type is void otherwise it might look like the called function is returning some kind of value.